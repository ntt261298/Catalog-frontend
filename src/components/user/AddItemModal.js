import React from 'react';
import { connect } from 'react-redux';
import {
  Button, Modal, ModalHeader, ModalBody, ModalFooter,
} from 'reactstrap';
import { PropTypes } from 'prop-types';
import { hideModal } from '../../actions/app';
import getCategories from '../../actions/category';
import { addItem } from '../../actions/item';


export class AddItemModal extends React.Component {
  state = {
    title: '',
    description: '',
    categoryID: '1',
  }

  componentDidMount() {
    this.props.getCategories();
  }

  onChangeInput = (e) => {
    this.setState({
      [e.target.name]: e.target.value,
    });
  }

  addItem = async () => {
    const { title, description, categoryID } = this.state;
    if (!title) {
      this.setState({
        message: 'Title must not be blank',
      });
      return;
    }
    if (!description) {
      this.setState({
        message: 'Description must not be blank',
      });
      return;
    }
    await this.props.addItem(title, description, categoryID, this.props.token);
    this.props.hideModal();
  }

  render() {
    const { modal, hideModal, category } = this.props;
    const { message } = this.state;
    const isAddItemModal = modal === 'addItem';
    return (
      <div>
        <Modal isOpen={isAddItemModal} toggle={hideModal}>
          <ModalHeader toggle={hideModal}>Add Item</ModalHeader>
          <ModalBody>
            <div>
              <label htmlFor="title">
                Title:
                <input type="text" name="title" onChange={this.onChangeInput} />
              </label>
            </div>
            <div>
              <label htmlFor="description">
                Description:
                <input type="text" name="description" onChange={this.onChangeInput} />
              </label>
            </div>
            <div>
              <label htmlFor="categoryID">
                Category:
                <select name="categoryID" onChange={this.onChangeInput}>
                  {
                        category.allIds.map(id => (
                          <option value={id}>{category.byId[id].name}</option>
                        ))
                    }
                </select>
              </label>
            </div>
            { message ? (
              <div className="alert alert-danger" role="alert">
                { message }
              </div>
            ) : null}
          </ModalBody>
          <ModalFooter>
            <Button color="primary" onClick={this.addItem}>Add</Button>
          </ModalFooter>
        </Modal>
      </div>
    );
  }
}

AddItemModal.propTypes = {
  modal: PropTypes.string.isRequired,
  token: PropTypes.string.isRequired,
  category: PropTypes.object.isRequired,
  hideModal: PropTypes.func.isRequired,
  addItem: PropTypes.func.isRequired,
  getCategories: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  modal: state.app.modal,
  token: state.user.accessToken,
  category: state.category,
});

export default connect(mapStateToProps, {
  getCategories, addItem, hideModal,
})(AddItemModal);
