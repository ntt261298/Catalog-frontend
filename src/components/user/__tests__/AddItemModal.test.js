import React from 'react';
import { shallow } from '../../../enzyme';
import { AddItemModal } from '../AddItemModal';

describe('src/components/user/AddItemModal', () => {
  let wrapper;
  let props;
  let addButton;

  const update = () => {
    addButton = wrapper.find('Button');
  };

  const setup = () => {
    wrapper = shallow(
      <AddItemModal {...props} />,
    );
    update();
  };

  beforeEach(() => {
    props = {
      category: {
        byId: {
          1: {
            id: 1,
            items: [],
            name: 'Category 1',
          },
          2: {
            id: 2,
            items: [],
            name: 'Category 2',
          },
        },
        allIds: [1, 2],
      },
      modal: 'addItem',
      token: 'randomToken',
      addItem: jest.fn(),
      hideModal: jest.fn(),
      getCategories: jest.fn(),
    };
  });

  it('should render add item modal', () => {
    setup();
    // Expect the wrapper object to be defined
    expect(wrapper).toMatchSnapshot();
  });

  it('should call addItem when add button is clicked', () => {
    setup();
    wrapper.setState({
      title: 'test', description: '123456', categoryID: 1,
    }, () => {
      addButton.simulate('click');
      expect(props.addItem).toHaveBeenCalled();
      expect(props.hideModal).toHaveBeenCalled();
    });
  });

  it('should call onChangeInput when user\'s input is changed', () => {
    setup();
    wrapper.setState({
      title: 'test', description: '123456', categoryID: 1,
    }, () => {
      addButton.simulate('click');
      expect(props.addItem).toHaveBeenCalled();
      expect(props.hideModal).toHaveBeenCalled();
    });
  });
});
