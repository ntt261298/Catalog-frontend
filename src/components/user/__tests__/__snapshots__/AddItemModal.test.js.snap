// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/user/AddItemModal should render add item modal 1`] = `
<div>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={true}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    role="dialog"
    toggle={[MockFunction]}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[MockFunction]}
      wrapTag="div"
    >
      Add Item
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <div>
        <label
          htmlFor="title"
        >
          Title:
          <input
            name="title"
            onChange={[Function]}
            type="text"
          />
        </label>
      </div>
      <div>
        <label
          htmlFor="description"
        >
          Description:
          <input
            name="description"
            onChange={[Function]}
            type="text"
          />
        </label>
      </div>
      <div>
        <label
          htmlFor="categoryID"
        >
          Category:
          <select
            name="categoryID"
            onChange={[Function]}
          >
            <option
              key="1"
              value={1}
            >
              Category 1
            </option>
            <option
              key="2"
              value={2}
            >
              Category 2
            </option>
          </select>
        </label>
      </div>
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="primary"
        onClick={[Function]}
        tag="button"
      >
        Add
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;
